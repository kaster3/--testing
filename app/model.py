from dataclasses import dataclass
from enum import Enum


class BookStatus(Enum):
    """
    Клас представляющий статус книг (в наличии, выдана), для более удобного взаимодействия
    и уменьшения шанса на ошибку
    """
    IN_STOCK = "В наличии"
    CHECKED_OUT = "Выдана"

    @staticmethod
    def from_value(value: str) -> "BookStatus":
        """
        Метод для получения статуса книги по его значению, для того чтобы json корректно работал с Enum
        """
        for status in BookStatus:
            if status.value == value:
                return status
        return BookStatus.IN_STOCK


@dataclass
class Book:
    """Основная бизнес сущность с которой будут выполняться все CRUD операции"""
    id: int
    title: str
    author: str
    year: int
    status: BookStatus = BookStatus.IN_STOCK


    def __str__(self):
        """
        Метод для вывода информации о книге в удобном виде
        """
        return (
            f"Книга №{self.id}\nНазвание: {self.title}\nАвтор:  {self.author}\n"
            f"Год издания: {self.year}\nСтатус выдачи: {self.status.value}\n"
        )


    def to_dict(self):
        """
        Метод для преобразования экземпляра книги в словарь так как json не умеет работать с классом Enum
        """
        return {
            'id': self.id,
            'title': self.title,
            'author': self.author,
            'year': self.year,
            'status': self.status.value
        }


    def __hash__(self):
        return hash((self.id, self.title, self.author, self.year, self.status))

    def __eq__(self, other):
        if not isinstance(other, Book):
            return NotImplemented

        return (self.id, self.title, self.author, self.year, self.status) == \
            (other.id, other.title, other.author, other.year, other.status)