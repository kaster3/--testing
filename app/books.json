[
    {
        "id": 1,
        "title": "Traveling",
        "author": "Shara",
        "year": 1999,
        "status": "В наличии"
    },
    {
        "id": 2,
        "title": "Traveling",
        "author": "John Smith",
        "year": 1999,
        "status": "В наличии"
    },
    {
        "id": 3,
        "title": "Terminator",
        "author": "Albec Ruso",
        "year": 1979,
        "status": "В наличии"
    }
]