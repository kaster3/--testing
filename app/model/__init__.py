__all__ = (
    "Task",
    "TaskPriority",
    "TaskStatus",
    "Category",
)

from .enums import Category, TaskPriority, TaskStatus
from .task import Task
