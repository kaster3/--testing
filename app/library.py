from .storage import Storage
from .model import Book, BookStatus


class LibraryCRUD:
    """
    Класс LibraryCRUD предназначен для прямого выполнения CRUD операций, а также работает
    с хранилищем книг, сохраняя туда данные и вытаскивая

    Args:
        self.storage: Класс хранилище
        self.books: Список книг, загруженных из json файла
    """
    def __init__(self, storage: Storage) -> None:
        self.storage: Storage = storage
        self.books: list[Book | None] = self.storage.load_books()


    def get_book_by_id(self, book_id: int | None) -> Book | None:
        """
        Метод для получения книги по ее id
        """
        for book in self.books:
            if book.id == book_id:
                return book


    def add_book(self, title: str, author: str, year: int) -> None:
        """
        Метод для добавления новой книги и ее сохранения на основе полученных параметров
        с автоматической генерации уникального id книги
        """
        # Генерация уникального id
        book_id: int = self.books[-1].id + 1 if self.books else 1
        new_book: Book = Book(book_id, title, author, year)
        self.books.append(new_book)
        self.storage.save_books(books=self.books)


    def remove_book(self, book: Book) -> None:
        """
        Метод удаления книги предварительно, которая была найдена
        """
        self.books.remove(book)
        self.storage.save_books(books=self.books)


    def get_books(
            self,
            title: str | None = None,
            author: str | None = None,
            year: int | None = None,
    ) -> set[Book] | list[Book | None]:
        """
        Метод получения, неповторяющихся, книг, по переданным параметрам
        """
        all_books: list[Book | None] = self.books
        result: set[Book | None] = set()

        if title:
            result.update([book for book in all_books if title.lower() in book.title.lower()])
        if author:
            result.update([book for book in all_books if author.lower() in book.author.lower()])
        if year:
            result.update([book for book in all_books if book.year == year])

        return result if result else all_books


    def change_status(self, book: Book) -> None:
        """
        Метод меняющий состояние статуса на противоположный у книги, которая была заранее найдена
        """
        if book.status == BookStatus.IN_STOCK:
            book.status = BookStatus.CHECKED_OUT
        else:
            book.status = BookStatus.IN_STOCK
        self.storage.save_books(books=self.books)
