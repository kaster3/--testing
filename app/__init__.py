__all__ = (
    "Storage",
    "Book",
    "BookStatus",
    "LibraryCRUD",
    "Service",
)

from .storage import Storage
from .model import Book, BookStatus
from .library import LibraryCRUD
from .service_layer import Service