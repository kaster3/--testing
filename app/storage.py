import json

from app.model import BookStatus, Book



class Storage:
    """
        Класс Storage предназначен для управления хранением и сохранением книг в формате JSON.

        Args:
            self.storage_file (str): имя файла для хранения книг
    """
    def __init__(self) -> None:
        self.storage_file = "books.json"

    def save_books(self, books) -> None:
        """
        Метод для сохранения списка книг в формате JSON в указанный файл или созднание этого файла при
        его отсутствие
        """
        with open(self.storage_file, 'w', encoding='utf-8') as file:
            json.dump([book.to_dict() for book in books], file, ensure_ascii=False, indent=4)
        print("Изменения сохранены в файл.\n")


    def load_books(self) -> list[Book | None]:
        """
        Метод для загрузки книг из файла JSON и преобразования их в список моделей Book
        """
        try:
            with open(self.storage_file, "r", encoding="utf-8") as file:
                books_data = json.load(file)

                # Преобразование словаря в модель Book и создание списка книг
                # Возможно выглядит сложно потому, что обязательно нужно привести статус в Enum
                unsorted_books: list[Book | None] = [
                    Book(
                        **{**data, "status": BookStatus.from_value(data["status"])}
                    )
                    for data in books_data
                ]
                # Сортировка книг по id, на случай, если книги в файле были переставлены,
                # так как будет генерироваться неправильный id, он генерируется: books[-1] + 1
                return sorted(unsorted_books, key=lambda book: book.id)

        except FileNotFoundError:
            return []
        except json.JSONDecodeError as exc:
            print(f"Ошибка при загрузке книг из файла: {exc}")
            return []
        except KeyError as exc:
            print(f"Ошибка преобразования статуса книги: {exc}")
            return []
