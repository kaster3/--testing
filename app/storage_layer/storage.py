import json

from app.messages.storage.consts import EMPTY_FILE, SAVE_CHANGES
from app.messages.storage.message_formatter import format_key_error_message
from app.model.task import Task
from app.utils.sort_tasks import sort_tasks


class TaskStorage:
    """
    Класс Storage предназначен для управления хранением и сохранением задач в формате JSON.

    Args:
        self.storage_file (str): имя файла для хранения задач
    """

    default_name = "storage_layer/tasks.json"

    def __init__(self, storage_file: str = default_name) -> None:
        self.storage_file = storage_file

    def save_task_json(self, tasks) -> None:
        """
        Метод для сохранения списка задач в формате JSON в указанный файл или созднание этого файла,
        при его отсутствие
        """
        with open(self.storage_file, "w", encoding="utf-8") as file:
            json.dump([task.to_dict() for task in tasks], file, ensure_ascii=False, indent=4)
        print(SAVE_CHANGES)

    def load_task_json(self) -> list[Task | None]:
        """
        Метод для загрузки задач из файла JSON и преобразования их в список моделей Task
        """
        try:
            with open(self.storage_file, "r", encoding="utf-8") as file:
                tasks_data = json.load(file)

                unsorted_tasks: list[Task | None] = [Task(**data) for data in tasks_data]

                # Сортируем по id, на случай, если в файле сделали перестановку книг, будет
                # неверно формироваться id новой книги в этом случает т.к.
                # правило list(Task...)[-1]["id]
                return sort_tasks(tasks=unsorted_tasks)

        except FileNotFoundError:
            return []
        except json.JSONDecodeError:
            print(EMPTY_FILE)
            return []
        except KeyError as exc:
            print(format_key_error_message(exc=exc))
            return []
