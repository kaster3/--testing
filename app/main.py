from app.messages.main.consts import MAIN_EXIT, WRONG_INPUT
from app.messages.main.message_formatter import main_exception
from app.data_accesor_layer.repository import TaskRepository
from app.service_layer import TaskService
from app.service_layer.service_helper import ServiceHelper
from app.storage_layer.storage import TaskStorage


def main(service: TaskService) -> None:
    """
    Главный цикл приложения, который обрабатывает взаимодействие с пользователем и вызывает
    соответствующие методы у класса TaskService, выполняя основную бизнес логику
    """

    choices = {
        "1": service.show_tasks,
        "2": service.add_task,
        "3": service.update_task,
        "4": service.delete_task,
        "5": service.search_by_key_word,
        "6": lambda: print(MAIN_EXIT),
    }

    while True:
        choice = service.get_choice()
        choices.get(choice, lambda: print(WRONG_INPUT))()
        if choice == "6":
            break


if __name__ == "__main__":
    storage = TaskStorage()
    repository = TaskRepository(storage=storage)
    service_helper = ServiceHelper()
    service = TaskService(repository=repository, service_helper=service_helper)
    try:
        main(service=service)
    except Exception as exc:
        print(main_exception(exc=exc))
