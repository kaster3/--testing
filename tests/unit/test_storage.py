import os

from syrupy import SnapshotAssertion

from app.model import Task
from app.storage_layer.storage import TaskStorage
from tests.messages.storage import TEST_FILE_NAME


class TestStorage:

    def test_save_task_json(self, created_tasks: list[Task], storage: TaskStorage):
        with open(storage.storage_file, "w", encoding="utf-8") as file:
            assert file.name == TEST_FILE_NAME
            assert os.path.exists(file.name) is True
            storage.save_task_json(tasks=created_tasks)

    def test_load_task_json(self, saved_file: TaskStorage, snapshot: SnapshotAssertion):
        loaded_tasks = saved_file.load_task_json()
        assert len(loaded_tasks) == 2
        snapshot.assert_match([task.to_dict() for task in loaded_tasks])
