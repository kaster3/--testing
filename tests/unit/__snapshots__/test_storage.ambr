# serializer version: 1
# name: TestStorage.test_load_task_json
  list([
    dict({
      'category': 'Личное',
      'description': 'test description 1',
      'due_date': '2026-12-20',
      'id': 1,
      'priority': 'Высокий',
      'status': 'Не выполнена',
      'title': 'test title 1',
    }),
    dict({
      'category': 'Работа',
      'description': 'test description 2',
      'due_date': '2025-12-23',
      'id': 2,
      'priority': 'Низкий',
      'status': 'Не выполнена',
      'title': 'test title 2',
    }),
  ])
# ---
